lib/email/message_id_service.rb

```diff
@@ -58,13 +62,54 @@ def generate_for_topic(topic, use_incoming_email_if_present: false, canonical: f
         end
       end
 
+      ##
+      # The outbound_message_id may be present because either:
+      #
+      # * The post was created via incoming email and Email::Receiver, and
+      #   references a Message-ID generated by an external email client or service.
+      # * At least one email has been sent because of the post being created
+      #   to inform interested parties via email.
+      #
+      # If it is blank then we should assume Discourse was the originator
+      # of the post, and generate a Message-ID to be used from now on using
+      # our discourse/post/POST_ID@HOST format.
+      def generate_or_use_existing(post)
+        if post.outbound_message_id.blank?
+          post.update(outbound_message_id: "discourse/post/#{post.id}@#{host}")
+        end
+        "<#{post.outbound_message_id}>"
+      end
+
+      def generate_or_use_existing_bulk(post_ids)
+        return [] if post_ids.empty?
+
+        DB.query_single(<<~SQL, host: host)
+          UPDATE posts
+          SET outbound_message_id = 'discourse/post/' || posts.id || '@' || :host
+          WHERE outbound_message_id IS NULL AND posts.id IN (#{post_ids.join(",")});
+          SELECT '<' || posts.outbound_message_id || '>'
+          FROM posts
+          WHERE posts.id IN (#{post_ids.join(",")})
+          ORDER BY posts.created_at ASC;
+        SQL
+      end
+
+      ##
+      # Uses extracted Message-IDs from both the In-Reply-To and References
+      # headers from an incoming email.
       def find_post_from_message_ids(message_ids)
         message_ids = message_ids.map { |message_id| message_id_clean(message_id) }
-        post_ids =  message_ids.map { |message_id| message_id[message_id_post_id_regexp, 1] }.compact.map(&:to_i)
-        post_ids << Post.where(
-          topic_id: message_ids.map { |message_id| message_id[message_id_topic_id_regexp, 1] }.compact,
-          post_number: 1
-        ).pluck(:id)
+
+        # TODO (martin) 2023-01-01 We should remove these backwards-compatible
+        # formats for the Message-ID and solely use the discourse/post/999@host
+        # format.
+        topic_ids = message_ids.map { |message_id| message_id[message_id_topic_id_regexp, 1] }.compact.map(&:to_i)
+        post_ids = message_ids.map { |message_id| message_id[message_id_post_id_regexp, 1] }.compact.map(&:to_i)
+
+        post_ids << message_ids.map { |message_id| message_id[message_id_discourse_regexp, 1] }.compact.map(&:to_i)
+
+        post_ids << Post.where(outbound_message_id: message_ids).pluck(:id)
+        post_ids << Post.where(topic_id: topic_ids, post_number: 1).pluck(:id)
```

Comment:
```suggestion
        post_ids << Post.where(outbound_message_id: message_ids).or(Post.where(topic_id: topic_ids, post_number: 1)).pluck(:id)
```

I think we can do this in a single query.